Add maybe function

Eruant opened the pull request:
I have not added documentation yet, as I want to make sure the function
API is correct first (feed back welcome)

Eruant committed:
Added maybe function

I have not added documentation yet, as I want to make sure the function
API is correct first (feed back welcome)

Eruant committed:
Merge branch 'master' into feature/maybe

skylize commented on lib/maybe.test.js:
```diff
@@ -0,0 +1,46 @@
+  t.is(maybe().open('isEmpty', x => x), 'isEmpty', 'should return empty option')
+})
+
+test('maybe.isNothing', t => {
```
Is `NaN` nothing?

Eruant replied:
From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NaN)...

> The global NaN property is a value representing Not-A-Number.

So it is technically a value. However it may be more useful to add a test such as `isNaN(contents)`

skylize replied:
Yeah. JS definitely considers it a value. I think there is a strong case for hiding that as a convenience and calling it a Nothing. Either way, I think a test for it is a good idea.

skylize replied:
I do not think we should bloat the codebase or api with NaN-checks if we decide to treat it as a value.

skylize commented on lib/maybe.mjs:
```diff
@@ -0,0 +1,26 @@
+  const map = fn =>
+    maybe(isNothing() ? null : fn(contents))
+
+  const open = curry(
```
`open` seems like the the wrong name here. I don't think I've seen that before. How about `fold`?

Eruant replied:
Yes I was wondering what a good name was for this. Lots of the inspiration for this came from [drboolean's mostly adequate guide](https://drboolean.gitbooks.io/mostly-adequate-guide/ch8.html#schr%C3%B6dingers-maybe), but with more functional programming.

In the guide it uses `Maybe` and `maybe`, which to me is confusing.

Eruant replied:
I think I may need to take a another look at the API. I'm just looking into Elm and how that uses `maybe`

skylize replied:
Look at my long comment below with Haskell examples. I think the naming conventions will start to make more sense. I can't specifically remember the Dr Boolean example at the moment, but I would say that `Maybe` is a type and `maybe` is a function that acts on a `Maybe`.

skylize replied:
Ahh yes. Here from your link, lowercase `maybe` is a helper, not implemented inside the `Maybe` constructor. 

> There is, however, an escape hatch. If we would rather return a custom value and continue on, we can use a little helper called maybe.
> ```js
> //  maybe :: b -> (a -> b) -> Maybe a -> b
> var maybe = curry(function(x, f, m) {
>   return m.isNothing() ? x : f(m.__value);
> });

skylize commented on lib/maybe.mjs:
```diff
@@ -0,0 +1,26 @@
+  const open = curry(
+    (isEmpty, fn) =>
+      isNothing()
+        ? isEmpty
```
Instead of an `isEmpty` value, should it call a provided function?

skylize replied:
I'm doing some research on maybes now, and I think a value is correct here, but we should name it something like `default` instead of `isEmpty`.

Eruant replied:
It should be clear to the user that this value will be sent back when `maybe` is not a value. Would `fallbackValue` work better?

skylize replied:
Yeah. I would even be down for a ridiculously long name here like `defaultValueForEmptyCase`.

skylize reviewed (changes requested):
This API is definitely incomplete, but we can always build on it

I wonder if we need `Maybe`. I was thinking about doing a `Right` and a `Left`. You can do a `fromNullable` implemented as basically a 1-liner from those, which gives pretty much the same functionality.

skylize commented:
Should we aim for Fantasyland on this one? I think this implementation is pretty far off from the spec.

skylize commented:
Ignoring Fantasyland for a moment, let's look at some Haskell for inspiration.

---
The `maybe` function.
(This is different from the `Maybe` type.)
```hs
maybe :: b -> (a -> b) -> Maybe a -> b

maybe False odd (Just 3) -- True
maybe False odd (Just 4) -- False
maybe False odd Nothing -- False
```

---
Some `Maybe` related type signatures for a few values.
```hs
Just "Hello" :: Maybe [Char]  -- reads as Just "Hello" is a Maybe of a list of characters
Just 42 :: (Num t) => Maybe t   -- reads as Just 42 is a Maybe of a value of type Num
Nothing :: Maybe a  -- reads as Nothing is a Maybe of a generic type
```

---
Some `Maybe` related expressions.
(The `::` operator is used here as typecasting.)
```hs
Just "Haha"  -- Just "Haha"  
Just 84   -- Just 84  
Just 10 :: Maybe Double  -- Just 10.0 
Nothing :: Maybe Double -- Nothing
```

---
A small sample of the `Maybe` type in use.
(You can read the `|` operators here like `if` / `else` and the `=` operators as `return`.)
```hs
safeLog :: (Floating a, Ord a) => a -> Maybe a
safeLog x
    | x > 0     = Just (log x)
    | otherwise = Nothing

safeLog 1000 -- Just 6.907755278982137
safeLog -1000 -- Nothing
```

---
So it looks to me like in an ideal implementation would provide the user with a `Just` constructor and an already constructed `Nothing` and some [helpful functions](https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Maybe.html) like `maybe`. The user would not necessarily use the `Maybe` constructor directly.

Some JS usage of this idealized implementation.
(This looks to me like it would be very pleasant to use.)
```js
import { Just, Nothing } from `lb-f`

const parse = str => {
  try { return Just(JSON.parse(str)) }
  catch (e) { return Nothing }
}

parse ('{a: 1}') // Just({ a: 1 })

maybe(parse('{b: 2}')) // { b: 2 }

parse ('random string') // Nothing

maybe(parse ('some text')) // Nothing
```

skylize commented:
I cannot think of a way to do type checking except either switching to TypeScript or using classes (es5 prototypes or sugar from es6 are both fine, just need `instanceof` to provide useful response). I think to build this properly, we need to do type checks in our tests.

(I guess you could argue that you aren't using classes just by using the `new` keyword. But seems to me by that point you are already on Class Road. You have inherited the downsides of classes, so you might as well reap the benefits.)

Eruant committed:
Merge branch 'master' into feature/maybe

Eruant committed:
Simplified the maybe functions

This maybe now fits more with the style of maybe in Elm

Eruant committed:
Merge branch 'feature/maybe' of github.com:littleball-games/lb-f into feature/maybe

Eruant committed:
Merge branch 'master' into feature/maybe

Eruant committed:
Updated maybe to use new API

Hopefully this API should be easier to use. I'm wondering if we should
set up a maybe function that we pass a maybe value to?

e.g.
```javascript
const resolveValue = maybe('no value', value => `got value "${value}"`)
const myMaybe = just('a value')

resolveValue(myMaybe) // got value "a value"
```

Eruant commented:
Hopefully this API should be easier to use. I'm wondering if we should
set up a maybe function that we pass a maybe value to?

e.g.
```javascript
const resolveValue = maybe('no value', value => `got value "${value}"`)
const myMaybe = just('a value')

resolveValue(myMaybe) // got value "a value"
```

skylize commented:
The simplicity is nice, but I think it misses the point. We should at least have a maybe that is a functor for composability with other functors, with our ideal target being to make it a proper monad. Your documentation case should work more like this.

```js

import { compose, just, nothing, pointfree as pf } from 'lb-f'
const map = pf('map')
const filter = pf('filter')

const users = [
  {
    name: 'Bob',
    nickname: 'Fluffy'
  },
  {
    name: 'Sarah'
  }
]

const greet = user => `Hello ${user.nickname}`

const maybeNickname = user =>
  user.nickname ? just(user) : nothing()

const safeGreet = compose(
  map(greet), // call the maybe's map method with greet function
  maybeNickname // get a maybe(nickname) from a user
)

const greetings = map(safeGreet, users) // [ just("Hello Fluffy"), nothing ]

// getOrElse is method on the maybe to extract internal value.
const get = pf('getOrElse')

compose(
  filter(x => x),  // remove false values
  map(get(false)) // map over array, extracting values or returning false
)(greetings)           // [ "Hello Fluffy" ]

// Because we maintained composability, we can still cut out the
// intermediate array if we already know we're ready to get at the values.
// Here instead of filtering, I've chosen an anonymous response

const getSafeGreeting = compose(
  get('Who are you?'),
  map(greet),
  maybeNickname
)
map(getSafeGreeting, users)                 // [ "Hello Fluffy", "Who are you?" ]

```

The people at [ramda-fantasy have pointed out](https://github.com/ramda/ramda-fantasy/issues/61) that the version of maybe in the Dr. Boolean book you are using as reference fails the functor test because of automatic conversion of null values to nothings. I am starting to agree with this. I think it should be up to the user to define what returns a nothing instead of the maybe deciding for them. His example was good for learning, but I think we can do better. [Folktale's Data.Maybe](http://folktalegithubio.readthedocs.io/en/latest/api/data/maybe/Maybe.html) looks to me like a better reference to work from. [folktale/data.maybe](https://github.com/folktale/data.maybe)

---
Also I'm a bit unclear why we're bothering to throw errors for only a small subset of invalid use cases. If user fails to pass 2 arguments, we throw with a custom message, but if either of those arguments is not a function, then it's going to crash anyways. So instead of bloating the code with custom errors, just let the stack tell them what's wrong.

Or else find a way to coerce the bad parameters into something that won't throw at all. (I prefer this solution if we can come up with something clever that user might actually want as a fallback.)

Eruant committed:
Updated maybe based on review feedback

Eruant commented:
> Also I'm a bit unclear why we're bothering to throw errors for only a small subset of invalid use cases. If user fails to pass 2 arguments, we throw with a custom message, but if either of those arguments is not a function, then it's going to crash anyways. So instead of bloating the code with custom errors, just let the stack tell them what's wrong.

I've removed this

> Or else find a way to coerce the bad parameters into something that won't throw at all. (I prefer this solution if we can come up with something clever that user might actually want as a fallback.)

I've set the a default error message. If something was not set correctly they should get this back.

Eruant committed:
Updated maybe functions

Added `maybe` as it's own function so that each function can be re-used.

Eruant commented:
I've moved away from Dr Boolean and have been looking at [elm's maybe](https://guide.elm-lang.org/error_handling/maybe.html) for reference.

I've changed the way `just` and `nothing` defined, and added a `maybe` function that acts as a resolver. This allows a set up function to be created that can be passed a _maybe_ object. The resulting code is now much simpler.

I've tried to avoid using other features such as `pointfree` in the docs to keep it simpler to understand (even if you would do that in practice).

skylize commented:
I like this. That fits with what I was describing about Haskell's implementation. Lowercase `maybe` is just a function that you pass a `Maybe` into. 👍 

We can start adding properties to `just` and `nothing` to make them into pointed functors and eventually monads.

Sure thing about not using `pointfree` in the docs for describing features. I was writing that comment to you, and I figured you could handle it. 😄 Also was good practice for me to use it as I keep trying to decide what is some good example code for `pointfree.docs.md`.

Eruant commented:
@skylize Great, are you able to approve the review (or do you not have enough access)?

skylize commented:
I don't like the use of `errorValue` parameter name. Maybes are often used for error handling, but a `Nothing` is not inherently an error, it's just a nothing.

Eruant commented:
> I don't like the use of errorValue parameter name. Maybes are often used for error handling, but a Nothing is not inherently an error, it's just a nothing.

Agreed, do you have any suggestions? `emptyValue`, `isNothing` or `nothingValue`?

I guess `successCallback` makes less sense in that case too.

skylize commented:
Best I've come up with so far is `elseValue`, but I'm not in love with it. Hmm...

I think we should play around a bit with adding more properties before committing it to master. This does the bare bare minimimum, but it's definitely incomplete. We need to be able to `map`, `chain`, etc. so let's make sure we aren't committing to a structure that will be painful to work with.

I suggest you rebase master into this branch, then split off from here onto another branch to continue working and leave this pull request in place for now.

I'm typing some code right now that I'm gonna request a pull onto this branch. I've already rebased master into my local copy, so you can safely force push the rebase onto this branch without messing me up.

```sh
git checkout feature/maybe
git rebase master
git push -f 
git checkout -b some-other-branch-name
```

Eruant committed:
Merge branch 'master' into feature/maybe

skylize commented:
just brainstorming here:
`valForNothing`, `justApplicator`

Eruant commented:
What about `nothingValue` and `justCallback` or `valueForNothing` and `callbackForJust`? (brainstorming too)

skylize commented:
`callbackForJust` vs `justApplicator`

I think that it better if we can avoid thinking of this as a callback, but rather as "function application". Definitely _not_ claiming that my brainstorm idea here is good. I'm just sharing the thought process that has me trying to steer away from a name that includes callback.

Here is the official description of haskell's maybe function that does the same thing.

> The maybe function takes a default value, a function, and a Maybe value. If the Maybe value is Nothing, the function returns the default value. Otherwise, it applies the function to the value inside the Just and returns the result.

Notice the phrase "applies the function to". This is a way of thinking that we should be trying to internalize and trying to steer others toward as well.

I definitely have a strong tendency to think of "_calling_ a function _with_ a parameter", but in trying to think functionally, it's better if we can transition that to "_applying_ a function _to_  a parameter"

Eruant commented:
Should we go for `defaultValue` and `justApplicator` then? I think I like that the best.

skylize commented:
I'm pretty happy with `defaultValue`, I think you were resistant to that in earlier discussions, but if you have come around, I'm in. 

Not excited by `justApplicator`, but I feel like it's sort of on the right track. What is an "applicator" anyway? Is that even a real word or did I just make it up? Haha

skylize commented:
It just came to me what an "applicator" is. It is a tool used to do application. So if anything, `maybe()` could be described as an applicator, but not so much the function provided as parameter. So `justApplicator` does not make any sense.
